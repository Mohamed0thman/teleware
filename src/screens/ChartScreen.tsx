import {StyleSheet, Text, View, TouchableOpacity} from 'react-native';
import React, {useState, useEffect, useRef} from 'react';
import RootScreen from '../component/RootScreen';
import {RootStackParamList} from '../navigation/RootNavigation';
import {StackScreenProps} from '@react-navigation/stack';
import {Currency} from '../models';
import {LineChart} from 'react-native-chart-kit';
import moment from 'moment';
import {getCurrencyRates} from '../store/slices/mainSlice';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {COLORS, FONTS, SCALE} from '../constants';
import {Spinner} from '../component/UI';

const {rh, rw, SC_Width} = SCALE;

type Props = StackScreenProps<RootStackParamList, 'Chart'> & {};

const ChartScreen = ({navigation, route}: Props) => {
  const {currency} = route.params as {currency: Currency};

  const [loading, setLoading] = useState<boolean>(false);

  const [date, setDate] = useState(moment().format('YYYY-MM-DD'));

  const {currencyRates} = useAppSelector(state => state.main);

  const dispatch = useAppDispatch();

  const subtractDay = moment().subtract(1, 'days').format('YYYY-MM-DD');
  const subtractMonth = moment().subtract(1, 'months').format('YYYY-MM-DD');
  const subtractYear = moment().subtract(1, 'years').format('YYYY-MM-DD');
  const subtractFiveYear = moment().subtract(5, 'years').format('YYYY-MM-DD');

  const filterDates = [
    {id: 1, date: subtractDay, title: '1D'},
    {id: 2, date: subtractMonth, title: '1M'},
    {id: 3, date: subtractYear, title: '1Y'},
    {id: 4, date: subtractFiveYear, title: '5Y'},
  ];

  function filterByDate(filterDate: string) {
    dispatch(getCurrencyRates({date: filterDate, base: currency.key}));
  }

  useEffect(() => {
    setLoading(true);
    dispatch(getCurrencyRates({date, base: currency.key})).finally(() =>
      setLoading(false),
    );
  }, [currency.key]);

  const renderFilter = () => {
    return (
      <View style={styles.filterContainer}>
        <Text style={styles.filterTitle}>Filtter</Text>
        <View style={styles.dates}>
          {filterDates.map((date, index) => (
            <TouchableOpacity
              key={`date${index}`}
              style={styles.date}
              onPress={() => {
                setDate(date.date);
                filterByDate(date.date);
              }}>
              <Text style={styles.dateText}>{date.title}</Text>
            </TouchableOpacity>
          ))}
        </View>
      </View>
    );
  };

  if (loading) return <Spinner />;
  return (
    <RootScreen containerStyle={styles.rootScreen}>
      {renderFilter()}
      {currencyRates && (
        <LineChart
          data={{
            labels: currencyRates?.keys || [''],

            datasets: [
              {
                data: currencyRates?.values || [1, 2, 3, 4],
              },
            ],
          }}
          width={SC_Width} // from react-native
          height={rh(600)}
          yAxisLabel="$"
          yAxisInterval={1} // optional, defaults to 1
          chartConfig={{
            backgroundColor: '#e26a00',
            backgroundGradientFrom: '#fb8c00',
            backgroundGradientTo: '#ffa726',
            decimalPlaces: 2, // optional, defaults to 2dp
            color: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
            labelColor: (opacity = 1) => `rgba(255, 255, 255, ${opacity})`,
            style: {
              borderRadius: 16,
            },
            propsForDots: {
              r: '6',
              strokeWidth: '2',
              stroke: '#ffa726',
            },
            barPercentage: 0.5,
            barRadius: 5,
          }}
          bezier
          style={{
            marginVertical: 8,
            borderRadius: 16,
          }}
          withInnerLines={false}
          withOuterLines={false}
        />
      )}
    </RootScreen>
  );
};

export default ChartScreen;

const styles = StyleSheet.create({
  rootScreen: {},
  filterContainer: {paddingHorizontal: rw(20), marginBottom: rh(20)},
  filterTitle: {...FONTS.h3, color: COLORS.black, marginVertical: rh(10)},
  dates: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    paddingHorizontal: rw(20),
  },
  date: {borderWidth: 1, borderColor: COLORS.black, padding: rw(20)},
  dateText: {...FONTS.h4, color: COLORS.black},
});
