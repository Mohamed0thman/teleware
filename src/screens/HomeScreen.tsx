import {
  StyleSheet,
  Text,
  View,
  FlatList,
  TouchableOpacity,
  RefreshControl,
} from 'react-native';
import React, {useEffect, useCallback} from 'react';
import RootScreen from '../component/RootScreen';
import {RootStackParamList} from '../navigation/RootNavigation';
import {StackScreenProps} from '@react-navigation/stack';
import {COLORS, FONTS, ICONS, SCALE, SIZES} from '../constants';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {getCurrencies} from '../store/slices/mainSlice';
import {Currency} from '../models';
import {SkeletonPlaceholder, Spinner, Typography} from '../component/UI';
import {useTranslation} from 'react-i18next';
import LinearGradient from 'react-native-linear-gradient';

const {rh, rw} = SCALE;
const {GearIcon} = ICONS;

type Props = StackScreenProps<RootStackParamList, 'Home'> & {};

const colors = [
  ['#a18cd1', '#fbc2eb'],
  ['#f9d423', '#ff4e50'],
  ['#6a11cb', '#2575fc'],
  ['#ff9a9e', '#fad0c4'],
];

const HomeScreen = ({navigation}: Props) => {
  const [refreshing, setRefreshing] = React.useState(false);
  const {isLoading, currencies, error} = useAppSelector(state => state.main);
  const {isRtl} = useAppSelector(state => state.setting);

  const {t} = useTranslation();
  const dispatch = useAppDispatch();

  const onRefresh = React.useCallback(() => {
    setRefreshing(true);
    dispatch(getCurrencies()).finally(() => setRefreshing(false));
  }, []);

  useEffect(() => {
    dispatch(getCurrencies());
  }, []);

  const renderItem = useCallback(
    ({item, index}: {item: Currency; index: number}) => {
      const backgroundColor = colors[index % colors.length];

      return (
        <TouchableOpacity
          testID="currencies"
          accessibilityLabel={`currency-${item.key}`}
          style={styles.currencie}
          onPress={() => navigation.navigate('Chart', {currency: item})}>
          <LinearGradient
            colors={backgroundColor}
            style={styles.linearGradient}>
            <Text testID={item.key} style={styles.key}>
              {item.key}
            </Text>
            <Text testID={item.name} style={styles.name}>
              {item.name}
            </Text>
            <Text testID={item.symbol} style={styles.symbol}>
              {item.symbol}
            </Text>
          </LinearGradient>
        </TouchableOpacity>
      );
    },
    [],
  );

  const renderSkeleton = () => {
    return (
      <View
      testID='skeletonPlaceHolder'
        style={[
          styles.currencie,
          {justifyContent: 'center', alignItems: 'center'},
        ]}>
        <SkeletonPlaceholder>
          <View style={styles.placholder}></View>
        </SkeletonPlaceholder>
      </View>
    );
  };

  // if (isLoading) return <Spinner />;

  console.log('isRtl', isRtl);

  return (
    <RootScreen containerStyle={styles.rootScreen}>
      <View
        style={{
          flexDirection: isRtl ? 'row' : 'row-reverse',
          justifyContent: 'space-between',
          alignItems: 'center',
        }}>
        <TouchableOpacity onPress={() => navigation.navigate('Settings')}>
          <GearIcon width={rw(24)} height={rh(24)} />
        </TouchableOpacity>

        <Typography testID="screen-title" isRtl={isRtl} style={styles.title}>
          {t('app:titles.HomeTitle')}
        </Typography>
      </View>
      {isLoading ? (
        <FlatList
          scrollEnabled
          horizontal={false}
          keyExtractor={(item, index) => `currencies-${index}`}
          showsVerticalScrollIndicator={false}
          numColumns={2}
          columnWrapperStyle={{justifyContent: 'space-between'}}
          data={new Array(12)}
          renderItem={renderSkeleton}
          ItemSeparatorComponent={() => <View style={{height: rh(20)}} />}
          refreshControl={
            <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
          }
        />
      ) : (
        <FlatList
          scrollEnabled
          horizontal={false}
          showsVerticalScrollIndicator={false}
          numColumns={2}
          keyExtractor={item => `currencies-${item.key}`}
          columnWrapperStyle={{justifyContent: 'space-between'}}
          data={currencies}
          contentContainerStyle={{paddingBottom: rh(20)}}
          renderItem={renderItem}
          ItemSeparatorComponent={() => <View style={{height: rh(20)}} />}
          refreshControl={
            <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
          }
        />
      )}
    </RootScreen>
  );
};

export default HomeScreen;

const styles = StyleSheet.create({
  rootScreen: {paddingHorizontal: rw(20)},
  currencie: {
    width: rw(150),
    height: rh(140),
    borderRadius: rw(SIZES.radius),
    overflow: 'hidden',
    shadowColor: '#000',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,

    elevation: 3,
  },
  linearGradient: {
    flex: 1,
    width: '100%',
    padding: rw(10),
    justifyContent: 'center',
    alignItems: 'center',
  },
  title: {...FONTS.h1, color: COLORS.black, marginVertical: rh(20)},
  key: {...FONTS.h3, color: COLORS.white},
  name: {...FONTS.h3, color: COLORS.white},
  symbol: {...FONTS.h3, color: COLORS.white},
  textholder: {
    height: rh(20),
    backgroundColor: COLORS.black,
    width: rw(50),
    marginTop: rh(5),
  },
  placholder: {
    backgroundColor: COLORS.black,
    flex: 1,
    width: '100%',
  },
});
