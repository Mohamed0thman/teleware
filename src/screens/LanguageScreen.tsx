import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  I18nManager,
} from 'react-native';
import React, {useEffect} from 'react';
import {useAppDispatch, useAppSelector} from '../store/configureStore';
import {useTranslation} from 'react-i18next';
import RootScreen from '../component/RootScreen';
import {COLORS, FONTS, ICONS, SCALE} from '../constants';
import {StackScreenProps} from '@react-navigation/stack';
import {RootStackParamList} from '../navigation/RootNavigation';
import {checkAndRequestNotificationPermission} from '../utils/permissions';
import {changeLang, setNotifee} from '../store/slices/settingSlice';
import {Typography} from '../component/UI';

const {rw, rh, SC_Width} = SCALE;
const {EgyptFlag, AmircaFlag} = ICONS;

type Props = StackScreenProps<RootStackParamList, 'Language'> & {};

const LanguageScreen = ({navigation}: Props) => {
  const {t, i18n} = useTranslation();
  const {isRtl} = useAppSelector(state => state.setting);
  const dispatch = useAppDispatch();

  const handleOnSelectLang = (lang: 'ar' | 'en', isRtl: boolean) => {
    i18n.changeLanguage(lang);
    dispatch(changeLang({lang, isRtl}));
    navigation.replace('Location');
  };

  useEffect(() => {
    checkAndRequestNotificationPermission().then(() =>
      dispatch(setNotifee(true)),
    );
  }, []);

  //   render button to select lang
  const langSelector = (lang: 'ar' | 'en', title: string, isRtl: boolean) => {
    return (
      <TouchableOpacity
        testID={`${lang}-btn`}
        style={styles.langButton}
        onPress={() => handleOnSelectLang(lang, isRtl)}>
        {/* icon */}
        {lang === 'ar' ? (
          <EgyptFlag width={rw(50)} height={rh(50)} />
        ) : (
          <AmircaFlag width={rw(50)} height={rh(50)} />
        )}
        <Text style={styles.langTitle}>{title}</Text>
      </TouchableOpacity>
    );
  };

  return (
    <RootScreen containerStyle={styles.rootScreen}>
      <Typography testID="screen-title" isRtl={isRtl} style={styles.title}>
        {t('app:titles.langScreen')}
      </Typography>
      <View
        style={[styles.btns, {flexDirection: isRtl ? 'row-reverse' : 'row'}]}>
        {langSelector('en', 'English', false)}
        {langSelector('ar', 'العربيه', true)}
      </View>
    </RootScreen>
  );
};

export default LanguageScreen;

const styles = StyleSheet.create({
  rootScreen: {},
  title: {
    ...FONTS.h2,
    color: COLORS.black,
    textAlign: 'center',
    marginTop: rh(40),
  },
  btns: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    gap: 50,
  },
  langButton: {
    justifyContent: 'center',
    alignItems: 'center',
    width: rw(80),
    height: rh(120),
    borderColor: COLORS.black,
    borderWidth: 1,
  },
  langTitle: {...FONTS.h3, color: COLORS.black},
});
