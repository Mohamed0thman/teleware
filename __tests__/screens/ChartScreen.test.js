import React from 'react';
import LanguageScreen from '../../src/screens/LanguageScreen';

import {fireEvent, screen} from '@testing-library/react-native';
import HomeScreen from '../../src/screens/HomeScreen';
import renderer, {act} from 'react-test-renderer';
import ChartScreen from '../../src/screens/ChartScreen';

describe('languge screen', () => {
  const route = {params: {currency: {key: 'EUR', name: 'Euro', symbol: 'â‚¬'}}};
  it('renders correctly', () => {
    renderWithRedux(<ChartScreen route={route} />);
  });

  it('renders correctly', () => {
    expect(renderWithRedux(<ChartScreen route={route} />)).toMatchSnapshot();
  });

  it('should render Spinner', async () => {
    // Wait for the component to be rendered

    const {getByTestId} = renderWithRedux(<ChartScreen route={route} />);

    const spinner = getByTestId('loading');

    expect(spinner).toBeDefined();
  });
});
