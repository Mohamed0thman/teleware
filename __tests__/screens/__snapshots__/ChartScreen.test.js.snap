// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`languge screen renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
  testID="loading"
>
  <ActivityIndicator
    color="#007864"
    size="large"
  />
</View>
`;
